### ISSUE TITLE
Wrong result type for || or ?? operation for interface value and object of some other type structure

### ISSUE THREAD

**Issue by blazkovicz** [microsoft/TypeScript#63018]
### 🔎 Search Terms

||
??
or
operator
typecheck

### 🕗 Version & Regression Information

typescript: 5.9.3

### 💻 Code

```ts
interface StrictInteface {
  value: string;
}

function someFunction(arg: StrictInteface) {
  return arg.value.trim();
}

const validValue: StrictInteface | null = null;

someFunction(validValue || { other: 125 });
someFunction(validValue ?? { other: 125 });

```

### 🙁 Actual behavior

No error shown by typecheck even though wrong value is given to the functions and they will fail at runtime.

### 🙂 Expected behavior

Both function calls should show type mismatch error

**Comment by MartinJohns** [comment]
You forgot to fill out the issue template, the playground link is missing.

In the [playground](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgMpisBYCS4LxLIDeAUMsgG5wA2ArhAFzIDOGoA5gNykC+ppGHRDZgAexCsxAWwgAxYaIkAKOFA7N0mbHkiEIAShLlkUCGDpRJajgDpq9CLfbTlBnv1IIJbKrWAAJgBqtAya7Dr4+sgAPsggdDQ0yAC88Yk0PKQsMvKKYOIgyg6BIY6xccTIYmAAFtDMAIwATACsyLzu2bkKIgUqJcGhKAD8IyTVdQ3ILe2dPEA) you can see that it does produce an error in both cases, so everything is working as it should.

So if you don't get any errors it's due to something in your setup.

**Comment by blazkovicz** [comment]
Wow, this is really interesting. I will check ts config. Thank you!

**Comment by blazkovicz** [comment]
Bug was in our code. All good. thanks.
